runtime/src/bank.rs

```diff
@@ -6534,14 +6544,58 @@ impl Bank {
         new_feature_activations: &HashSet<Pubkey>,
     ) {
         for builtin in BUILTINS.iter() {
+            // The `builtin_is_bpf` flag is used to handle the case where a
+            // builtin is scheduled to be enabled by one feature gate and
+            // later migrated to Core BPF by another.
+            //
+            // There should never be a case where a builtin is set to be
+            // migrated to Core BPF and is also set to be enabled on feature
+            // activation on the same feature gate. However, the
+            // `builtin_is_bpf` flag will handle this case as well, electing
+            // to first attempt the migration to Core BPF.
+            //
+            // The migration to Core BPF will fail gracefully because the
+            // program account will not exist. The builtin will subsequently
+            // be enabled, but it will never be migrated to Core BPF.
+            //
+            // Using the same feature gate for both enabling and migrating a
+            // builtin to Core BPF should be strictly avoided.
```

Comment:
This is certainly possible to support, although I question whether or not this makes sense at all, as I mentioned to @Lichtso [here](https://github.com/anza-xyz/agave/pull/590#discussion_r1578740743).

My thinking here is that it doesn't make sense to activate a builtin and then migrate it to BPF in the same epoch, when you can just deploy it directly to BPF.

If a feature-gated BPF deployment, in general, is something you're after, I think that's best handled beyond this PR. It could be useful for upgrades.

However, if you think the ability to mistakenly add the same feature gate for both is too brittle, perhaps we can remove the `enable_feature_id` altogether, as long as we're sure we'll never add more builtins.

What do you think? If you have other ideas on why we'd like to enable and migrate in the same epoch, maybe elaborate?

Reply:
I don't think we will add more built-ins, everything should be deployed as core program directly. So enabling and migrating won't happen either.

Reply:
We can't remove the `enable_feature_id` until ZK Token and Loader V4 features are activated. Going off the schedule, it appears ZK Token is blocked and the other isn't scheduled yet.

In the meantime, though, maybe I can just document the `enable_feature_id` field with a comment? Or perhaps I can deprecate it, but I don't think that will be as obvious as we think, since it's internal.

Reply:
@CriesofCarrots another point - re: your first suggestion here - is that enabling builtins is idempotent unless `only_apply_transitions_for_new_features` is `false`. It's always `true` on epoch rollovers, ie. `Caller::NewFromParent`.

Since migrations are a one-off, we have to check that first in order to skip the otherwise idempotent enabling. This is another reason for that `builtin_is_bpf` check I've included, as well as the test case you were [questioning here](https://github.com/anza-xyz/agave/pull/590#discussion_r1581595117).